File: src/DesignPatterns/src/C++/002-GOFPatterns/003-BehavioralPatterns/009-Visitor/ComputerVisitor/Computer.cpp
#include "Computer.h"
#include "ComputerPartVisitor.h"

void Computer::AddPart(IComputerPart *pc)
{
	m_parts.push_back(pc);
}
void Computer::Accept(IComputerVisitor *pcv)
{
	for (auto cp : m_parts)
		cp->Accept(pcv);

	pcv->Visit(this);
}

File: src/DesignPatterns/src/Java/002-GOF-Patterns/003-BehavioralPatterns/009-Visitor/Visitor/src/org/csystem/app/Computer.java
package org.csystem.app;

import java.util.ArrayList;
import java.util.List;

public class Computer implements IComputerPart {
    private final List<IComputerPart> m_computerPartList = new ArrayList<>();

    //...

    public boolean add(IComputerPart iComputerPart)
    {
        //...
        return m_computerPartList.add(iComputerPart);
    }

    @Override
    public void accept(IComputerVisitor visitor)
    {
        m_computerPartList.forEach(cp -> cp.accept(visitor));
        visitor.visit(this);
    }
}
